hub/core/chunk_engine.py

```diff
@@ -0,0 +1,326 @@
+        return self.chunk_id_encoder.num_samples
+
+    @property
+    def last_chunk(self) -> Optional[Chunk]:
+        if self.num_chunks == 0:
+            return None
+
+        last_chunk_name = self.chunk_id_encoder.get_name_for_chunk(-1)
+        last_chunk_key = get_chunk_key(self.key, last_chunk_name)
+        return self.cache.get_cachable(last_chunk_key, Chunk)
+
+    @property
+    def tensor_meta(self):
+        tensor_meta_key = get_tensor_meta_key(self.key)
+        return self.cache.get_cachable(tensor_meta_key, TensorMeta)
+
+    def _append_bytes(
+        self, incoming_buffer: memoryview, shape: Tuple[int], dtype: np.dtype
+    ):
+        """Treat `incoming_buffer` as a single sample and place them into `Chunk`s. This function implements the algorithm for
+        determining which chunks contain which parts of `incoming_buffer`.
+
+        Args:
+            incoming_buffer (memoryview): Buffer that represents a single sample that may or may not be compressed.
+            shape (Tuple[int]): Shape for the sample that `incoming_buffer` represents.
+            dtype (np.dtype): Data type for the sample that `incoming_buffer` represents.
+        """
+
+        num_samples = 1
+        incoming_num_bytes = len(incoming_buffer)
+
+        last_chunk = self.last_chunk or self._create_new_chunk()
+
+        # update tensor meta first because erroneous meta information is better than un-accounted for data.
+        self.tensor_meta.check_compatibility(shape, dtype)
+        self.tensor_meta.update(shape, dtype, num_samples)
+
+        # TODO: turn into more functions
+
+        last_chunk_extended = False
+        forwarding_buffer = incoming_buffer
+        if last_chunk.is_under_min_space(self.min_chunk_size_target):
+            last_chunk_size = last_chunk.num_data_bytes
+            chunk_ct_content = _min_chunk_ct_for_data_size(
+                self.max_chunk_size, incoming_num_bytes
+            )
+
+            extra_bytes = min(incoming_num_bytes, self.max_chunk_size - last_chunk_size)
+            combined_chunk_ct = _min_chunk_ct_for_data_size(
+                self.max_chunk_size, incoming_num_bytes + last_chunk_size
+            )
+
+            # combine if count is same
+            if combined_chunk_ct == chunk_ct_content:
+                last_chunk.append_sample(
+                    forwarding_buffer[:extra_bytes], self.max_chunk_size
+                )
+                forwarding_buffer = forwarding_buffer[extra_bytes:]
+                self._synchronize_last_chunk(num_samples, incoming_num_bytes, shape)
+                last_chunk_extended = True
+
+        # check if `last_chunk_extended` to handle empty samples
+        if len(forwarding_buffer) > 0 or not last_chunk_extended:
+            new_chunk = self._create_new_chunk()
+            new_chunk.append_sample(forwarding_buffer, self.max_chunk_size)
+            self._synchronize_last_chunk(num_samples, incoming_num_bytes, shape)
```

Comment:
:man_shrugging: if it works it works, I guess

Reply:
it's clean now